src/main/scala/dsptools/numbers/UIntTypeclasses.scala

```diff
@@ -3,99 +3,116 @@
-      println("Saturating subtract is broken right now!")
-      val grow = f -& g
-      val nogrow = f -% g
-      val max = 3.U
-      val min = 0.U
-
-      Mux(grow === nogrow,
-        nogrow,
-        Mux(grow > 0.U, max, min)
-      )
-    } else {
-      f -% g
+    val diff = {
+      if(context.overflowType == Grow) f -& g
+      else if (context.overflowType == Wrap) f -% g
+      else throw DspException("Saturating subtractor hasn't been implemented")
     }
+    ShiftRegister(diff, context.numAddPipes)
   }
-}
-
-trait UIntImpl {
-  implicit object UIntIntegerImpl extends UIntInteger
-  implicit def fromInt(x: Double): UInt = x.toInt.U
+  def negate(f: UInt): UInt = throw DspException("Can't negate UInt and get UInt")
+  def times(f: UInt, g: UInt): UInt = {
+    ShiftRegister(f * g, context.numMulPipes)
+  }  
 }
 
 trait UIntOrder extends Any with Order[UInt] with hasContext {
   override def compare(x: UInt, y: UInt): ComparisonBundle = {
     ComparisonHelper(x === y, x < y)
   }
+  override def eqv(x: UInt, y: UInt): Bool = x === y
+  override def neqv(x: UInt, y:UInt): Bool = x =/= y
+  override def lt(x: UInt, y: UInt): Bool = x < y
+  override def lteqv(x: UInt, y: UInt): Bool = x <= y
+  override def gt(x: UInt, y: UInt): Bool = x > y
+  override def gteqv(x: UInt, y: UInt): Bool = x >= y
+  // min, max depends on lt, gt
 }
 
 trait UIntSigned extends Any with Signed[UInt] with hasContext {
   def signum(a: UInt): ComparisonBundle = {
     ComparisonHelper(a === 0.U, a < 0.U)
   }
-
-  /** An idempotent function that ensures an object has a non-negative sign. */
   def abs(a: UInt): UInt = a // UInts are unsigned!
+  override def isSignZero(a: UInt): Bool = a === 0.U
+  override def isSignPositive(a: UInt): Bool = !isSignZero(a)
+  override def isSignNegative(a: UInt): Bool = false.B
+  // isSignNonZero, isSignNonPositive, isSignNonNegative derived from above (!)
 }
+
 trait UIntIsReal extends Any with IsIntegral[UInt] with UIntOrder with UIntSigned with hasContext {
   def toDouble(a: UInt): DspReal = ???
+  // In IsIntegral: ceil, floor, round, truncate (from IsReal) already defined as itself; 
+  // isWhole always true
+  
+  def isOdd(a: UInt): Bool = a(0)
+  // isEven derived from isOdd
+  // Note: whatever Chisel does -- is there overflow?
+  // Generally better to use your own mod if you know input bounds
+  def mod(a: UInt, b: UInt): UInt = a % b
 }
 
 trait ConvertableToUInt extends ConvertableTo[UInt] with hasContext {
   def fromShort(n: Short): UInt = n.toInt.U
-  //def fromAlgebraic(n: Algebraic): UInt = n.toBigInt.U
   def fromBigInt(n: BigInt): UInt = n.U
   def fromByte(n: Byte): UInt = n.toInt.U
-  def fromDouble(n: Double): UInt = n.toInt.U
-//  def fromReal(n: Real): UInt = n.toInt.U
-  //def fromRational(n: Rational): UInt = n.toInt.U
-  def fromType[B](n: B)(implicit c: ConvertableFrom[B]): UInt = c.toBigInt(n).U
+  def fromDouble(n: Double): UInt = n.round.toInt.U
```

Comment:
Is it better to round here? Do you have a specific case in mind where it is bad not to round?